### PR TITLE
Features/language switcher behavior on welcome pages #154187395

### PR DESCRIPTION
**Pull request by callaghanc**

### CODE REVIEW

File: spec/javascripts/services/shared_service_spec.coffee [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      $window = _$window_
+      $document = _$document_
+      SharedService = _SharedService_
+      return
+    )
+
+    describe 'getLanguageCode', ->
+      it 'returns the 2-letter code for the given language name', ->
+        code = SharedService.getLanguageCode('Spanish')
+        expect(code).toEqual 'es'
```
Comment (creativecoder) [inline spec/javascripts/services/shared_service_spec.coffee]: @callaghanc Yay, `SharedService` tests! What about testing all languages by adding the rest to `languageMap` and then
```coffee
Object.keys(languageMap).forEach key ->
  language = languageMap[key]
  languageCode = SharedService.getLanguageCode(language)
  expect(languageCode).toEqual key
```

File: app/assets/javascripts/shared/SharedService.js.coffee [hunk 2]
```diff
@@ -18,6 +18,40 @@ SharedService = ($http, $state, $window, $document) ->
...
+  Service.languageMap =
+    en: 'English'
+    es: 'Spanish'
+    tl: 'Filipino'
+    zh: 'Chinese'
+
+  Service.getLanguageCode = (langName) ->
+    # will take "English" and return "en", for example
+    if langName
+      capitalizedLangName = langName.charAt(0).toUpperCase() + langName.slice(1)
```
Comment (creativecoder) [inline app/assets/javascripts/shared/SharedService.js.coffee]: `_.startCase` also works for this.

File: app/assets/javascripts/shared/components/languageSwitcher.js.coffee [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
-      if $state.current.name == 'dahlia.welcome'
-        href = href + "/"
-      href
+      toStateName = $state.current.name
+
+      if SharedService.isWelcomePage($state.current)
+        if lang == 'en'
+          toStateName = 'dahlia.welcome'
+        else
+          longLang = SharedService.languageMap[lang].toLowerCase()
```
Comment (creativecoder) [inline app/assets/javascripts/shared/components/languageSwitcher.js.coffee]: Use `SharedService.getLanguageName`?

**Comment by creativecoder** [comment]
@callaghanc See above for a few code suggestions.

Also, I can't get to any of the language welcome pages in the review app--all have js stuck in an infinite loop ("call stack size exceeded").

File: app/assets/javascripts/shared/SharedService.js.coffee (hunk 2 above)
Comment (callaghanc) [reply app/assets/javascripts/shared/SharedService.js.coffee]: 🙏 yes thank you i was casting about for something like that, much preferable

File: app/assets/javascripts/shared/components/languageSwitcher.js.coffee (hunk 3 above)
Comment (callaghanc) [reply app/assets/javascripts/shared/components/languageSwitcher.js.coffee]: yep 👍

File: spec/javascripts/services/shared_service_spec.coffee (hunk 1 above)
Comment (callaghanc) [reply spec/javascripts/services/shared_service_spec.coffee]: sure, added testing of all languages to the language name and code function tests

**Comment by callaghanc** [comment]
@creativecoder made some code updates, looks like it is working again on review app, this should be good to go

**Pull request merged**